### PR TITLE
Brightcove Prebid Plugin ver 0.1.1

### PR DESCRIPTION
**Pull request by ynisman**
1. Specify adServerCallback by name
2. Playlisting support
3. Playlisting frequency control
4. VAST Wrapper limit support
5. Bug fixes during testing cycle

### CODE REVIEW

File: src/VastManager.js [hunk 1]
```diff
@@ -56,6 +61,102 @@ var vastManager = function () {
...
+					_playlistCreative = creative;
+					_player.playlist.autoadvance(!!_playlistCreative ? null : 0);
+				});
+			}
+		}
+		else {
+			_player.off('playlistitem', nextListItemHandler);
+		}
+	}
+
```
Comment (vpitts) [inline src/VastManager.js]: Would be helpful to add comments for each of the new functions

Comment (ynisman) [reply src/VastManager.js]: Added a few comments.

File: src/PrebidCommunicator.js [hunk 2]
```diff
@@ -47,28 +60,41 @@ var prebidCommunicator = function () {
...
+							func = _options.adServerCallback;
+						}
+						else if (typeof _options.adServerCallback === 'string' &&
+								 window[_options.adServerCallback] &&
+								 typeof window[_options.adServerCallback] === 'function') {
+							func = window[_options.adServerCallback];
+						}
+						if (func) {
+							func(arrBids, function(adServerCreative) {
+								_logger.log(_prefix, 'Selected VAST url: ' + creative);
```
[the remark this answers is not in the document]
Comment (ynisman) [reply src/PrebidCommunicator.js]: Fixed.

File: tests/e2e/testPages/prebid-body-playlist.html [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+</div>
+
+<div id='dlgOptions' style='display: none; position: absolute; top: 0px; left: 100px; z-index: 20000; background-color: orange; border:solid 1px black;'>
+	<p id='title'><b>&nbsp&nbsp&nbsp&nbspBrightcove VMAP plugin Options</b></p>
+    <table>
+    	<tr>
+    		<td>
+	    		<div>
+				    <p><b>Options</b></p>
+				    <input id="txtOptions" size="350" value='' style='display:none;'></input>
```
[the remark this answers is not in the document]
Comment (ynisman) [reply tests/e2e/testPages/prebid-body-playlist.html]: Removed unneeded closing tag

File: src/BcPrebidVast.js [hunk 4]
```diff
@@ -75,14 +79,24 @@ function dispatchPrebidDoneEvent() {
 
 // the function loads prebid.js and does header bidding if needed
```
Comment (vpitts) [inline src/BcPrebidVast.js]: This function is sufficient complicated and also very important to do prebid in the header.  I think it needs a little more comments to explain all the steps.  In particular, what is the meaning of the fromHeader argument?  How is that set?  What kind of argument is it?  Also how does loadPrebidScript get invoked?

File: src/BcPrebidVast.js [hunk 5]
```diff
@@ -73,16 +77,34 @@ function dispatchPrebidDoneEvent() {
 	document.dispatchEvent(event);
 }
 
-// the function loads prebid.js and does header bidding if needed
+// This function will be invoked immediately when plugin script is loaded in the page or by brightcove player.
```
Comment (mkendall07) [inline src/BcPrebidVast.js]: consider using JSdocs for documenting function signatures. Take a look at prebid.js for some examples https://github.com/prebid/Prebid.js/blob/master/src/prebid.js#L86

File: src/BcPrebidVast.js [hunk 6]
```diff
@@ -73,16 +77,34 @@ function dispatchPrebidDoneEvent() {
...
+			// invoke prebid
     		doPrebid(options, function(bids) {
-    			var arrBids = (bids && bids[options.biddersSpec.code]) ? bids[options.biddersSpec.code].bids : [];
+				// this function returns creative with higher CPM
+				var selectWinnerByCPM = function(arrBids) {
+					var cpm = 0.0;
+					var creative = null;
+					for (var i = 0; i < arrBids.length; i++) {
+						if (arrBids[i].cpm > cpm) {
+							cpm = arrBids[i].cpm;
```
Comment (mkendall07) [inline src/BcPrebidVast.js]: you might consider adjusting spacing to 2 spaces instead of 4 for deeply nested code readability.

Comment (ynisman) [reply src/BcPrebidVast.js]: My code editor uses tab to control indentation and the size of a tab is aet globally to 4. We find that using 4 spaces for indentation makes the code more readable in general. Changing indentation spacing in particular areas will be inconvenient.

**Pull request merged**